### ISSUE TITLE
bdev_nvme: Leak of an abort bdev_io when device admin queue is full

### ISSUE THREAD

**Issue by mhae** [spdk/spdk#3035]
# Sighting report

I've noticed that we leak/lose an abort bdev_io for an admin command when the admin queue of the NVMe device is full. The stack is NVMF/bdev_nvme.
When a NVMe command can't be placed on the admin queue, then the ENOMEM is returned. This error handling is not implemented for aborts of admin commands.

The culprit is here (SPDK 21.04)
```
static int
bdev_nvme_abort(struct nvme_io_channel *nvme_ch, struct nvme_bdev_io *bio,
		struct nvme_bdev_io *bio_to_abort)
{
	int rc;

	bio->orig_thread = spdk_get_thread();

	rc = spdk_nvme_ctrlr_cmd_abort_ext(nvme_ch->ctrlr->ctrlr,
					   nvme_ch->qpair,
					   bio_to_abort,
					   bdev_nvme_abort_done, bio);
	if (rc == -ENOENT) {
		/* If no command was found in I/O qpair, the target command may be
		 * admin command. Only a single thread tries aborting admin command
		 * to clean I/O flow.
		 */
		spdk_thread_send_msg(nvme_ch->ctrlr->thread,
				     bdev_nvme_abort_admin_cmd, bio);
		rc = 0;
	}

	return rc;
}
```
When we go into `bdev_nvme_abort_admin_cmd` and `spdk_nvme_ctrlr_cmd_abort_ext` fails with ENOMEM, then we simply return the function. There is no completion of the abort for ENOMEM or returning status back to the bdev layer.

This is in contrast to aborting non-admin commands where we don't go into bdev_nvme_abort_admin_cmd. Here we return ENOMEM to bdev layer which then puts the bdev_io on the nomem_io queue, keeping track of this IO.

Now, when we want to hot remove this device, we are unable to because of the outstanding/lost IO.

We probably want to simply complete the abort when we hit the NOMEM condition.

We also need to double-check the 23.05 code where the bdev_nvme_abort function no longer returns anything. Are we handling the nomem condition properly?
```
case SPDK_BDEV_IO_TYPE_ABORT:
		nbdev_io->io_path = NULL;
		nbdev_io_to_abort = (struct nvme_bdev_io *)bdev_io->u.abort.bio_to_abort->driver_ctx;
		bdev_nvme_abort(nbdev_ch,
				nbdev_io,
				nbdev_io_to_abort);
```

## Expected Behavior

I can hot remove a device, even when the admin queue is full.

## Context (Environment including OS version, SPDK version, etc.)
22.04, Debian 11.

**Comment by shuhei-matsumoto** [comment]
Hi @mhae,

I think a recently merged patch https://github.com/spdk/spdk/commit/f319b98d40c4ef1c45bdd2d39bbe8ea2ed5f10f2 should fix this issue.
Will you take a look? Should we backport this patch to v23.01.x?

**Comment by mhae** [comment]
@shuhei-matsumoto , I'll take a look.

**Comment by mhae** [comment]
@shuhei-matsumoto, I looked at the patch you've mentioned above. Wouldn't that be quite a lift to back port it due to the differences between 23.01 and main?

Could we try to find a smaller fix for 23.01.x? I would think that we could either expand the error check in bdev_nvme_abort_admin_cmd to
```
// if (rc == -ENOENT) {
	if (rc) {
		/* If no admin command was found in admin qpair or we encountered an error (e.g. NO MEM), complete the abort
		 * request with failure.
		 */
		bio->cpl.cdw0 |= 1U;
		bio->cpl.status.sc = SPDK_NVME_SC_SUCCESS;
		bio->cpl.status.sct = SPDK_NVME_SCT_GENERIC;

		spdk_thread_send_msg(bio->orig_thread, bdev_nvme_abort_completion, bio);
	}
```

Or check specifically for ENOMEM and do a
```
spdk_bdev_io_complete(bdev_io, SPDK_BDEV_IO_STATUS_NOMEM);
```
The latter would require and additional context parameter for `bdev_nvme_abort_completion` to pass the EMOMEM or we create a new bdev_nvme_abort_completion_non_mem function that just does the SPDK_BDEV_IO_STATUS_NOMEM complete. 

I'm currently doing some testing with the first approach but also want to try out the 2nd approach.

**Comment by shuhei-matsumoto** [comment]
@mhae I like your 1st approach. Will you submit a patch to the v23.01.x?

**Comment by mhae** [comment]
Super. Thanks for confirming. I'll do the patch for 23.01.x.

**Comment by shuhei-matsumoto** [comment]
@mhae,

Thank you for the update. I was confused and missed that. ENOENT indicated the target I/O was completed. In this case, the NVMe driver did not submit any abort command. Hence, the NVMe bdev module completed the abort command immediately with error.

**Comment by mhae** [comment]
@tomzawadzki , 23.01 is fine. The issue exists in 22.04. I can't remember if 22.04 is LTS but the issue is somewhat nasty. I'm fine with closing it or providing a fix for 22.XX LTS, although it would be nice to have a record of this issue in case someone else runs into it.

**Comment by tomzawadzki** [comment]
Let's close this issue since the issue was fixed by 23.01 and we are now at 24.01+.
